orca-webhook/src/main/java/com/netflix/spinnaker/orca/webhook/pipeline/PreconfiguredWebhookStage.java

```diff
@@ -0,0 +1,103 @@
+import com.netflix.spinnaker.orca.webhook.service.WebhookService;
+import com.netflix.spinnaker.orca.webhook.tasks.MonitorWebhookTask;
+import com.netflix.spinnaker.security.AuthenticatedRequest;
+import java.lang.reflect.Field;
+import java.util.Arrays;
+import java.util.List;
+import java.util.Map;
+import java.util.Objects;
+import java.util.stream.Collectors;
+import javax.annotation.Nonnull;
+import lombok.extern.slf4j.Slf4j;
+import org.springframework.beans.factory.annotation.Autowired;
+import org.springframework.stereotype.Component;
+
+@Component
+@Slf4j
+public class PreconfiguredWebhookStage extends WebhookStage {
+
+  private static List<String> IGNORE_FIELDS =
+      List.of("props", "enabled", "label", "description", "type", "parameters");
+  private static List<Field> ALL_FIELDS =
+      Arrays.stream(PreconfiguredWebhook.class.getDeclaredFields())
+          .filter(f -> !f.isSynthetic())
+          .filter(f -> !IGNORE_FIELDS.contains(f.getName()))
+          .collect(Collectors.toList());
+
+  private final FiatService fiatService;
+  private final WebhookService webhookService;
+
+  @Autowired
+  PreconfiguredWebhookStage(
+      WebhookService webhookService,
+      FiatService fiatService,
+      MonitorWebhookTask monitorWebhookTask) {
+    super(monitorWebhookTask);
+
+    this.webhookService = webhookService;
+    this.fiatService = fiatService;
+  }
+
+  @Override
+  public void taskGraph(@Nonnull StageExecution stage, @Nonnull TaskNode.Builder builder) {
+    var preconfiguredWebhook =
+        webhookService.getPreconfiguredWebhooks().stream()
+            .filter(webhook -> Objects.equals(stage.getType(), webhook.getType()))
+            .findFirst()
+            .orElseThrow(() -> new PreconfiguredWebhookNotFoundException(stage.getType()));
+
+    var permissions = preconfiguredWebhook.getPermissions();
+    if (permissions != null && !permissions.isEmpty()) {
+      String user = AuthenticatedRequest.getSpinnakerUser().orElse("anonymous");
+      var userPermission = fiatService.getUserPermission(user);
+
+      boolean isAllowed = preconfiguredWebhook.isAllowed("WRITE", userPermission.getRoles());
+      if (!isAllowed) {
+        throw new PreconfiguredWebhookUnauthorizedException(user, stage.getType());
+      }
+    }
+
+    overrideIfNotSetInContextAndOverrideDefault(stage.getContext(), preconfiguredWebhook);
+    super.taskGraph(stage, builder);
+  }
+
+  /** Mutates the context map. */
+  private static void overrideIfNotSetInContextAndOverrideDefault(
+      Map<String, Object> context, PreconfiguredWebhook preconfiguredWebhook) {
+    ALL_FIELDS.forEach(
+        it -> {
+          try {
+            if (context.get(it.getName()) == null || it.get(preconfiguredWebhook) != null) {
+              context.put(it.getName(), it.get(preconfiguredWebhook));
+            }
+          } catch (IllegalAccessException e) {
+            log.warn("unexpected reflection issue for field '{}'", it.getName());
```

Comment:
maybe worth just throwing here - would be in line with what would have happened in the original groovy code I think?

Reply:
updated